File: problems/LeetCode/tree/connect.py
def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
		if not root: return None

		q = deque([root])

		while q:
				l = len(q)
				for i in range(l):
						node = q.popleft()

						node.next = q[0] if i < l - 1 else None

						if node.left: q.append(node.left)
						if node.right: q.append(node.right)
						
		return root



def connect2(self, root: 'Optional[Node]') -> 'Optional[Node]':
		node = root

		while node:
				# next line nodes
				nxt = dummy = Node()
				while node:
						if node.left:
								nxt.next = node.left
								nxt = node.left
						if node.right:
								nxt.next = node.right
								nxt = node.right
						node = node.next
				# move to next line
				node = dummy.next
		
		return root

File: problems/LeetCode/TopInterview150/connect.ts
function connect(root: _Node | null): _Node | null {
  if (!root) return root;

  let queue = [];
  queue.push(root);

  while (queue.length) {
    const tmp = [];

    for (let i = 0; i < queue.length; i++) {
      if (i + 1 < queue.length) {
        queue[i].next = queue[i + 1];
      }

      if (queue[i].left) {
        tmp.push(queue[i].left);
      }

      if (queue[i].right) {
        tmp.push(queue[i].right);
      }
    }

    queue = [...tmp];
  }
  return root;
}

function connect2(root: _Node | null): _Node | null {
  if (!root) return root;

  let queue = [];
  queue.push(root);

  while (queue.length) {
    const len = queue.length;

    for (let i = 0; i < len; i++) {
      if (i + 1 < len) {
        queue[i].next = queue[i + 1];
      }

      if (queue[i].left) {
        queue.push(queue[i].left);
      }

      if (queue[i].right) {
        queue.push(queue[i].right);
      }
    }

    queue = queue.slice(len, queue.length);
  }
  return root;
}

// Time complexity: O(n)
// Space compleixity: O(n)

function connect3(root: _Node | null): _Node | null {
  if (!root) return root;

  let cur = root;

  while (cur) {
    const dummy = new _Node();
    let nxt = dummy;

    while (cur) {
      if (cur.left) {
        nxt.next = cur.left;
        nxt = nxt.next;
      }

      if (cur.right) {
        nxt.next = cur.right;
        nxt = nxt.next;
      }
      // move to the right subtree
      cur = cur.next;
    }
    // move to the next level
    cur = dummy.next;
  }
  return root;
}

// Time complexity: O(n)
// Space compleixity: O(1)
